### ISSUE TITLE
Allow lazy initialization logic with Standalone Components

### ISSUE THREAD

**Issue by yjaaidi** [angular/angular#45970]
### Which @angular/* package(s) are relevant/related to the feature request?

core

### Description

In general, some components need some logic to be initialized globally. (e.g. : Ngrx's state & effects).

This initialization logic can be added:
1. at bootstrap with `.forRoot()` pattern
2. lazily by importing a module with the `.forChild()` or `.forFeature()` pattern. _(e.g.  `EffectsModule.forFeature(...)`)_

With standalone components and without considering the NgModule compatibility function `importProvidersFrom`, we now have 2 main options for triggering initialization logic:

1. using `APP_INITIALIZER` at bootstrap,
```ts
bootstrapApplication(AppComponent, {
  providers: [
    {
      provide: APP_INITIALIZER,
      multi: true,
      useValue() {
        console.log('init...');
        inject(Store).registerEffects([...]);
      },
    },
  ],
});
```

2. using `ENVIRONMENT_INITIALIZER` in router providers.
```ts
{
  path: '...',
  providers: [
    {
      provide: ENVIRONMENT_INITIALIZER,
      multi: true,
      useValue() {
        console.log('init...');
        inject(Store).registerEffects([...]);
      },
    }
  ],
  children: [
    ...
  ],
}
```

The problem with the first approach (i.e. bootstrap) is the eager loading of all the logic (e.g. states / effects) and this doesn't scale as the main bundle will keep getting larger and larger.
The second option (`ENVIRONMENT_INITIALIZER` with router providers) allows lazy loading but is it the highest common point after the bootstrap? Not really. In fact, a state will mostly be used by components under some shared route but quickly while the app grows, one might want to reuse the component under another route or anywhere else in the app and that’s when things get tricky. As we broke the locality principle, we have to think about checking out the route providers and move the state/effects to somewhere higher… which will end up to be the bootstrap… and thus the clutter in the main bundle.
In addition to this, we can mention the challenge of debugging the state/effects provided in the routing. Someone tries to use a component outside the route and they observe the following surprising behavior “if I navigate to the feature route first, then I move to the other route, the component works, but when I land directly on this route, the action doesn’t change the state” (because the effect wasn’t loaded).

**TL;DR: providing state at bootstrap clutters the bundle and providing state at router level doesn’t scale either as one has to keep thinking about which state should be provided**

### Concrete example:
An app with an UnreadEmailComponent and bunch of other components used in /emails . Email state/effects are provided at the route level.
Months/years later, someone wants to lazy load and dynamically inject UnreadEmailComponent in a dialog or something like that when the user clicks somewhere in the landing page.
If they visit /emails first, it’s going to work but before navigating to /emails the effects/state are not initialized. First, they end up with a bug, then they need to be quite familiar with Ngrx and injectors to understand what’s happening and that’s when they  decide to move the state/effects to bootstrap which is annoying because we end up eager loading state/effects while the components that need them are lazy loaded.

### `LAZY_APP_INITIALIZER` injection token

- if provided at bootstrap, it would work like APP_INITIALIZER,
- if provided with the routing or dynamic injection, it should be called once (we shouldn’t run a function if the same reference is provided again somewhere else) when the first depending environment is created,
- if provided at node (i.e. component/directive/pipe) level, it should be called once if not already triggered during bootstrap, or env creation, or another component.

By the way, what would be the side effects of making this the default behavior of `APP_INITIALIZER`. It’s a breaking change, but what kind of apps/libs have lazy imported APP_INITIALIZER providers that are not executed and that would break the app if executed? Is that a plausible scenario?

This would allow something like this:
```ts
function provideEffects(effects) {
  return {
    provide: LAZY_APP_INITIALIZER,
    multi: true,
    useValue() {
      inject(Store).registerEffects(effects);
    }
  }
}

const emailFeatureProviders = [provideEffects([...]), provideState(...), EmailFacade];

@Component({
  standalone: true,
  template: ...,
  providers: [emailFeatureProviders]
})
export class UnreadEmail {} 

@Component({
  standalone: true,
  template: ...,
  providers: [emailFeatureProviders],
})
export class EmailInfo {} 
```

### Alternatives considered

@alxhub's side-effect trick:

```ts
function provideStoreWithFeature(feature: ...): Provider[] {
  return [
    {
      provide: Store,
      useFactory: () => {
        const rootStore = inject(RootStore);
        for (const feature of inject(STORE_FEATURE, [])) {
          rootStore.registerFeature(feature);
        }
        return rootStore;
      },
    }, {
      provide: STORE_FEATURE,
      useValue: feature,
      multi: true,
    },
  ];
}
```

**Comment by hiepxanh** [comment]
"main bundle will keep getting larger and larger." correct. I have the same situation, exactly as you described

**Comment by Den-dp** [comment]
I wonder, is there an option to address both this and `MODULE_INITIALIZER` https://github.com/angular/angular/issues/17606 requests?

I see how asynchronous providers https://github.com/angular/angular/issues/23279 can solve this but [DI is the wrong layer to address async dependencies](https://github.com/angular/angular/issues/23279#issuecomment-1165030809)

**Comment by jordanpowell88** [comment]
We too am running into this issue with Cypress Component Testing. Currently we are unable to access an injected NgRx `Store` instances without exposing it inside the component itself. Not sure if this feature is on the roadmap or not but personally would love to see something like this put in place.

**Comment by mauriziocescon** [comment]
Indeed, managing reusable components powered by ngrx-store requires modules right now... at least I'm not aware of any clean replacement of: 

``` typescript
import * as effects from './store/my_reusable_standalone_component.effects';
import { feature } from './store/my_reusable_standalone_component.feature';

@NgModule({
    imports: [
        StoreModule.forFeature(feature),
        EffectsModule.forFeature(effects),

        MyReusableStandaloneComponentUsingNgrxStore,
    ],
    exports: [
        MyReusableStandaloneComponentUsingNgrxStore,
    ],
})
export class MyReusableStandaloneComponentUsingNgrxStoreModule {
}
```

MyReusableStandaloneComponentUsingNgrxStore can be imported by other Standalone components / modules and everything is fine. 

Indeed a store-feature-factory at component level would be a great improvement. 

``` typescript
@Component({
  standalone: true,
  providers: [provideEffectsFactory(effects), provideStateFactory(feature)],
})
export class MyReusableStandaloneComponentUsingNgrxStore {}
```

Not a big deal in the end... but maybe @brandonroberts has a better view on the topic!

**Comment by mlc-mlapis** [comment]
Hmm, I've also been thinking for a long time about a new addition to a standalone component level that would allow initializing NgRx states and effects on the global scope, which would be equivalent to the old NgModule providers.

What about adding something like `rootProviders` in metadata of a component, like:
```
@Component({
   standalone: true,
   selector: mySelector,
   ...
   imports: [
      ...
   ],
   rootProviders: [
      provideState(...),
      provideEffects(...)
   ]
})
```
which would be an equivalent to the old NgModule syntax:
```
@NgModule({
    imports: [
        StoreModule.forFeature(feature),
        EffectsModule.forFeature(effects)
    ]
})
```
and also to the route syntax:
```
export default [
  {
    path: '',
    component: myPage,
    providers: [
      provideState(...),
      provideEffects(...)
    ]
  }
];
```
even in cases when no routes are used or necessary at all?

**Comment by brandonroberts** [comment]
> Indeed, managing reusable components powered by ngrx-store require modules right now... at least I'm not aware of any clean replacement of:
> 
> ```
> import * as effects from './store/my_reusable_standalone_component.effects';
> import { feature } from './store/my_reusable_standalone_component.feature';
> 
> @NgModule({
>     imports: [
>         StoreModule.forFeature(feature),
>         EffectsModule.forFeature(effects),
> 
>         MyReusableStandaloneComponentUsingNgrxStore,
>     ],
>     exports: [
>         MyReusableStandaloneComponentUsingNgrxStore,
>     ],
> })
> export class MyReusableStandaloneComponentUsingNgrxStoreModule {
> }
> ```
> 
> MyReusableStandaloneComponentUsingNgrxStore can be imported by other Standalone components / modules and everything is fine.
> 
> Personally, I find the addition of a store-feature at component.providers level the only (currently) available way to go, but kind of not right: it's "an addition at global level (a new slice on the app-store)" rather than at component one ([created / destroyed based on the component life cycle](https://angular.dev/guide/ngmodules/providers#limiting-provider-scope-with-components)).
> 
> ```
> @Component({
>   standalone: true,
>   providers: [provideEffects(effects), provideState(feature)],
> })
> export class MyReusableStandaloneComponentUsingNgrxStore {}
> ```
> 
> One might say that the registration at routes level is basically the same... https://ngrx.io/guide/store/reducers#using-the-standalone-api-1
> 
> Indeed... and again I have the feeling it's breaking the scope idea of providers registration. Not a big deal in the end... but maybe @brandonroberts has a better view on the topic!

We brought this up before also when the `ENVIRONMENT_INITIALIZER` was introduced, and the response was that a Component/NodeInjector would not receive the same type of initializer token to use at the component level. You could wire the feature state and effects up manually, but you also wouldn't get the automatic cleanup as expected in a component. 

As mentioned, NgRx Store features/effects are meant to be global/long-lived states and are not bound to individual component lifecycles. The recommended option for local component state is to use NgRx ComponentStore in the Component.providers where its tied to the lifecycle of the component.

**Comment by brandonroberts** [comment]
> Hmm, I've also been thinking for a long time about a new addition to a standalone component level that would allow initializing NgRx states and effects on the global scope, which would be equivalent to the old NgModule providers.
> 
> What about adding something like `rootProviders` in metadata of a component, like:
> 
> ```
> @Component({
>    standalone: true,
>    selector: mySelector,
>    ...
>    imports: [
>       ...
>    ],
>    rootProviders: [
>       provideState(...),
>       provideEffects(...)
>    ]
> })
> ```
> 
> which would be an equivalent to the old NgModule syntax:
> 
> ```
> @NgModule({
>     imports: [
>         StoreModule.forFeature(feature),
>         EffectsModule.forFeature(effects)
>     ]
> })
> ```
> 
> and also to the route syntax:
> 
> ```
> export default [
>   {
>     path: '',
>     component: myPage,
>     providers: [
>       provideState(...),
>       provideEffects(...)
>     ]
>   }
> ];
> ```
> 
> even in cases when no routes are used or necessary at all?

As mentioned, you can do this today with NgRx Store/Effects APIs, but you'd still have to provide/inject some arbitrary service to initialize the state/effects being registered.

**Comment by mlc-mlapis** [comment]
> As mentioned, you can do this today with NgRx Store/Effects APIs, but you'd still have to provide/inject some arbitrary service to initialize the state/effects being registered.

Sure, but there is also a factor of encapsulation that would allow you to write a bit more readable code if such a possibility would be available.

**Comment by mauriziocescon** [comment]
Say, IMO the global nature of a store.feature (not linked to the lifecycle of a component) is fine (an added value). 

I'm mostly thinking about a reusable component that needs to be rendered several times in several contexts... and for which you wanna reuse data (stateful case). 

``` typescript
// typical ng pure implementation 

@Injectable({ providedIn: 'root' })
export class MyDataService {

    // - created the first time it's used; 
    // - powers one or several instances of MyReusableStandaloneComponent
    // - tree shakable
    // - no need to bother AppModule or equivalent
}

@Component({ ..., standalone: true, ...})
export class MyReusableStandaloneComponent { 

    private myDataService = inject(MyDataService); 
    ... 
}

// just import the component where it's needed
```

``` typescript
// typical ngrx-store implementation with modules (mostly equivalent of the above one)

@Component({ ..., standalone: true, ...})
export class MyReusableStandaloneComponent { ... }

@NgModule({
    imports: [
        StoreModule.forFeature(feature),
        EffectsModule.forFeature(effects),

        MyReusableStandaloneComponent,
    ],
    exports: [
        MyReusableStandaloneComponent,
    ],
})
export class MyReusableStandaloneComponentModule {
}

// just import the module where it's needed
```

@brandonroberts I wonder: I assume there isn't currently a clean (and equivalent) way of achieving the ngrx-store result above without modules, right?

Assuming the answer is yes :-), we go back to the point: would `Component.providers[provideState(...)]` be a good replacement of `NgModule.imports[StoreModule.forFeature(feature)]`? 

Still think it's not ideal and we're missing something new at ng level...

**Comment by mlc-mlapis** [comment]
@mauriziocescon It can't be the `Component.providers[...]` because it's already used name with different scope. Now you need to create your own `provideXXX` function, like: 
```
export const provideXXX = (...): EnvironmentProviders => {
  return makeEnvironmentProviders([
    provideState(...),
    provideEffects(...)
  ]);
}
```
and call it somewhere.

**Comment by mauriziocescon** [comment]
@mlc-mlapis Sorry, it wasn't clear: my bad! 

Well, I was assuming that with the introduction of `LAZY_APP_INITIALIZER` (or similar) the factories could be moved to the ngrx-store side (not immediately of course). And the pattern becomes some sort of standard (like forRoot / forChild). 

Mine was mostly: should providers be used or something else (like you suggested)?

**Comment by TomaszStanik** [comment]
any update?

**Comment by inmativ** [comment]
I agree with the question. We have a complex project with a federation of modules. We use dynamic standaloone components and render them via ViewContainerRef.createComponent(). We need to register the localization for ngx-translate the first time we use the component, but there may be other similar scenarios. The root module does not know anything about these components, it is inappropriate there. Routing is also not used when rendering them.

Now we do this:

```typescript
const STANDALONE_INITIALIZER = new InjectionToken<void>('STANDALONE_INITIALIZER', {
  providedIn: 'root',
  factory: () => {
    const translate = inject(TranslateService);
    translate.setTranslation(...);
  },
});

@Component({...})
export class MyDinamicComponent {
  constructor() {
    inject(STANDALONE_INITIALIZER)
  }
}
``` 

But I would like to get rid of this crutch.
